homeassistant/components/stt/__init__.py

```diff
@@ -0,0 +1,217 @@
+
+    setup_tasks = [
+        async_setup_platform(p_type, p_config)
+        for p_type, p_config in config_per_platform(config, DOMAIN)
+    ]
+
+    if setup_tasks:
+        await asyncio.wait(setup_tasks)
+
+    hass.http.register_view(SpeechToTextView(providers))
+    return True
+
+
+@attr.s
+class SpeechMetadata:
+    """Metadata of audio stream."""
+
+    language: str = attr.ib()
+    format: AudioFormats = attr.ib()
+    codec: AudioCodecs = attr.ib()
+    bit_rate: AudioBitRates = attr.ib(converter=int)
+    sample_rate: AudioSampleRates = attr.ib(converter=int)
+
+
+@attr.s
+class SpeechResult:
+    """Result of audio Speech."""
+
+    text: str = attr.ib()
+    result: SpeechResultState = attr.ib()
+
+
+class Provider(ABC):
+    """Represent a single STT provider."""
+
+    hass: Optional[HomeAssistantType] = None
+    name: Optional[str] = None
+
+    @property
+    @abstractmethod
+    def supported_languages(self) -> List[str]:
+        """Return a list of supported languages."""
+
+    @property
+    @abstractmethod
+    def supported_formats(self) -> List[AudioFormats]:
+        """Return a list of supported formats."""
+
+    @property
+    @abstractmethod
+    def supported_codecs(self) -> List[AudioCodecs]:
+        """Return a list of supported codecs."""
+
+    @property
+    @abstractmethod
+    def supported_bit_rates(self) -> List[AudioBitRates]:
+        """Return a list of supported bit rates."""
+
+    @property
+    @abstractmethod
+    def supported_sample_rates(self) -> List[AudioSampleRates]:
+        """Return a list of supported sample rates."""
+
+    @abstractmethod
+    async def async_process_audio_stream(
+        self, metadata: SpeechMetadata, stream: StreamReader
+    ) -> SpeechResult:
+        """Process an audio stream to STT service.
+
+        Only streaming of content are allow!
+        """
+
+    @callback
+    def check_metadata(self, metadata: SpeechMetadata) -> bool:
+        """Check if given metadata supported by this provider."""
+        if (
+            metadata.language not in self.supported_languages
+            or metadata.format not in self.supported_formats
+            or metadata.codec not in self.supported_codecs
+            or metadata.bitrate not in self.supported_bit_rates
+            or metadata.samplerate not in self.supported_sample_rates
+        ):
+            return False
+        return True
+
+
+class SpeechToTextView(HomeAssistantView):
+    """STT view to generate a text from audio stream."""
+
+    requires_auth = True
+    url = "/api/stt/{provider}"
+    name = "api:stt:provider"
+
+    def __init__(self, providers: Dict[str, Provider]) -> None:
+        """Initialize a tts view."""
+        self.providers = providers
+
+    @staticmethod
+    def _metadata_from_header(request: web.Request) -> Optional[SpeechMetadata]:
+        """Extract metadata from header.
+
+        X-Speech-Content: format=wav; codec=pcm; samplerate=16000; bitrate=16; language=de_de
```

Comment:
We need to change samplerate and bitrate to sample_rate and bit_rate here too I think.